framework/sandstone.cpp

```diff
@@ -3563,6 +3574,22 @@ int main(int argc, char **argv)
                 sApp->shmem->max_messages_per_thread = INT_MAX;
             break;
 
+        case vary_frequency:
+#ifdef _WIN32
+            fprintf(stderr, "--vary-frequency doesn't work on windows\n");
+            return EXIT_FAILURE;
+#else
+            if (getuid() != 0) {
+                fprintf(stderr, "--vary-frequency requires user to be root\n");
+                return EXIT_NOPERMISSION;
+            }
+            sApp->vary_frequency_mode = true;
+
+            //before running the test, do a initial setup that also saves initial state for all cpus
+            sApp->frequency_manager.initial_setup();
```

Comment:
I also recommend you **don't** `getuid()` (because that's the wrong system call anyway, you actually meant `geteuid()`). Instead, do the normal initialisation; when you get to the point of setting the governor to "userspace", you'll open a file in sysfs for writing and THAT will tell you if you are allowed to and if those files exist. That is better because:
* avoids TOCTOU problems
* avoids using proxy information for what you really needed to know
* catches more situations where "--vary-frequency" may fail

Just make sure the error messages are understandable by the user.